lib/python/numpy.h

```diff
@@ -85,54 +76,144 @@ auto cast_to_array_like(const py::object &obj, const units::Unit unit) {
+            for (scipp::index k = 0; k < src.shape(2); ++k)
+              for (scipp::index l = 0; l < src.shape(3); ++l, ++it)
+                copy_element<convert>(src(i, j, k, l), *it);
       });
 }
 
-template <bool convert, class T, class View>
-void copy_flattened_3d(const py::array_t<T> &data, View &&view) {
-  auto r = data.unchecked();
-  auto it = view.begin();
-  for (scipp::index i = 0; i < r.shape(0); ++i)
-    for (scipp::index j = 0; j < r.shape(1); ++j)
-      for (scipp::index k = 0; k < r.shape(2); ++k, ++it)
-        copy_element<convert>(r(i, j, k), *it);
+template <bool convert, class T, class Dst>
+void copy_flattened_5d(const py::array_t<T> &src_array, Dst &dst) {
+  const auto src = src_array.template unchecked<5>();
+  const auto begin = dst.begin();
+  core::parallel::parallel_for(
+      core::parallel::blocked_range(0, src.shape(0), 10000),
+      [&](const auto &range) {
+        auto it = begin + range.begin() * src.shape(1) * src.shape(2) *
+                              src.shape(3) * src.shape(4);
+        for (scipp::index i = range.begin(); i < range.end(); ++i)
+          for (scipp::index j = 0; j < src.shape(1); ++j)
+            for (scipp::index k = 0; k < src.shape(2); ++k)
+              for (scipp::index l = 0; l < src.shape(3); ++l)
+                for (scipp::index m = 0; m < src.shape(4); ++m, ++it)
+                  copy_element<convert>(src(i, j, k, l, m), *it);
+      });
 }
 
-template <bool convert, class T, class View>
-void copy_flattened_4d(const py::array_t<T> &data, View &&view) {
-  auto r = data.unchecked();
-  auto it = view.begin();
-  for (scipp::index i = 0; i < r.shape(0); ++i)
-    for (scipp::index j = 0; j < r.shape(1); ++j)
-      for (scipp::index k = 0; k < r.shape(2); ++k)
-        for (scipp::index l = 0; l < r.shape(3); ++l, ++it)
-          copy_element<convert>(r(i, j, k, l), *it);
+template <bool convert, class T, class Dst>
+void copy_flattened_6d(const py::array_t<T> &src_array, Dst &dst) {
+  const auto src = src_array.template unchecked<6>();
+  const auto begin = dst.begin();
+  core::parallel::parallel_for(
+      core::parallel::blocked_range(0, src.shape(0), 10000),
+      [&](const auto &range) {
+        auto it = begin + range.begin() * src.shape(1) * src.shape(2) *
+                              src.shape(3) * src.shape(4) * src.shape(5);
+        for (scipp::index i = range.begin(); i < range.end(); ++i)
+          for (scipp::index j = 0; j < src.shape(1); ++j)
+            for (scipp::index k = 0; k < src.shape(2); ++k)
+              for (scipp::index l = 0; l < src.shape(3); ++l)
+                for (scipp::index m = 0; m < src.shape(4); ++m)
+                  for (scipp::index n = 0; n < src.shape(5); ++n, ++it)
+                    copy_element<convert>(src(i, j, k, l, m, n), *it);
+      });
+}
+
+template <bool convert, class T, class Dst>
+void copy_flattened(const py::array_t<T> &src_array, Dst &dst) {
```

Comment:
What is the difference between this and the `1d` function? Can one use the other?

Reply:
The 1d function works for any strides by indexing the array directly. The flattened function 'indexes' a pointer and only works for c-contiguous data.